Add _WIN64 preprocessor to use _ftelli64 and _fseeki64

lambertlulala opened the pull request:
Add `_ftelli64` and `_fseeki64` to fix https://github.com/universal-ctags/libreadtags/issues/36.

Note to properly use 8-byte `off_t`, the header `config.h` is included on Windows. Since `_mingw_off_t.h` (it is included in `stdio.h`)  is defined as:

``` C
#ifndef _OFF_T_DEFINED
#define _OFF_T_DEFINED
#ifndef _OFF_T_
#define _OFF_T_
  typedef long _off_t;
#if !defined(NO_OLDNAMES) || defined(_POSIX)
  typedef long off32_t;
#endif
#endif

#ifndef _OFF64_T_DEFINED
#define _OFF64_T_DEFINED
  __MINGW_EXTENSION typedef long long _off64_t;
#if !defined(NO_OLDNAMES) || defined(_POSIX)
  __MINGW_EXTENSION typedef long long off64_t;
#endif
#endif /*_OFF64_T_DEFINED */


#ifndef _FILE_OFFSET_BITS_SET_OFFT
#define _FILE_OFFSET_BITS_SET_OFFT
#if !defined(NO_OLDNAMES) || defined(_POSIX)
#if (defined(_FILE_OFFSET_BITS) && (_FILE_OFFSET_BITS == 64))
typedef off64_t off_t;
#else
typedef off32_t off_t;
#endif /* #if !defined(NO_OLDNAMES) || defined(_POSIX) */
#endif /* (defined(_FILE_OFFSET_BITS) && (_FILE_OFFSET_BITS == 64)) */
#endif /* _FILE_OFFSET_BITS_SET_OFFT */


#endif /* _OFF_T_DEFINED */

```

Note this PR is only tested on Windows and built from mingw64.

lambertlulala committed:
Add _WIN32 preprocessor to use _ftelli64 and _fseeki64

masatake commented:
I'll add more platforms as the test environments before merging this.

masatake commented on readtags.c:
```diff
@@ -9,6 +9,9 @@
 *   INCLUDE FILES
 */
+#ifdef _WIN64
+#include <config.h>  /* to define _FILE_OFFSET_BITS for Windows */
```
Can we expect config.h is always available on the platform where _WIN64 is defined?

masatake commented on readtags.c:
```diff
@@ -285,7 +288,11 @@ static int readTagLineRaw (tagFile *const file, int *err)
 		file->pos = ftell (file->fp);
+#else
+		file->pos = _ftelli64 (file->fp);
+#endif
```
Instead of putting #ifndef/#else/#endif multiple places, we can introduce functions like
```
static off_t readtags_ftell(FILE *fp);
static int   readtags_fseek(FILE *fp, off_t pos, int whence);
```

We can aggregate the changes into the implementation of the two functions.

What do you think about this idea?

k-takata commented:
Whether `_ftelli64` and `_fseeki64` are available or not doesn't depend on the definition of `_WIN64`. It only depends on the version of the C runtime library.

masatake commented:
@k-takata, do you mean we should check the availability of the functions with configure?

k-takata commented:
In my understanding, the functions were introduced in VC++ 2005. Currently, u-ctags uses VC++ 2013 or later, so I don't think we need to check the existence of them.

For mingw, mingw-w64 supports them, but the original version of mingw may not supports them. I'm not sure if we need to care about the original mingw.

lambertlulala commented:
Hello @masatake , I think that's great to implement those two functions, it's more elegant. For the `config.h`, I'm not sure whether it exists on other platforms such as Linux and Mac. Hence, I add _WIN64.

Hello @k-takata , based on my understanding, both `_ftelli64 `and `_fseeki64` are provided by Windows. The reason to use `_WIN64` is that ctags is a cross-platform program and that definition is for Windows only. The other choice I Know is _WIN32, however, `_WIN32` is to have libreadtags compiled on both 32-bit `and` 64-bit Windows. To compile it `only for 64-bit Windows`, _WIN64 is used.

Currently, I can't figure out other elegant way to handle it. But don't mind refactoring it if there is any good idea. I just throw a brick to attract jade.

k-takata commented:
For such purpose, `_WIN32` is suitable. The ...i64 functions are also available on 32-bit Windows and `_WIN32` is defined both on 32- and 64-bit Windows.

lambertlulala commented:
@k-takata You are right. `_WIN32` is enough and compatible for 32-bit and 64-bit. I have updated the commit and added those 2 functions suggested by @masatake .

masatake commented:
@k-takata 
> For mingw, mingw-w64 supports them, but the original version of mingw may not supports them. I'm not sure if we need to care about the original MinGW.

Is "the original version of mingw" equal to "MSYS2 / testing (MINGW32)" ?

I introduced "MSYS2 / testing (MINGW32)" as a test platform without understanding well.

The building on the platform failed because config.h was not found.

@lambertlulala, thank you for updating your pull request.

I got a question when I remembered your original pull request.

As far s I can remember, you used '0LL' instead of 'OL' at the place calling `_fseeki64`.
I'm thinking about what we should do about 0, 0L, and 0LL. If you know what we should do, could you reflect your idea on this pull request?

lambertlulala committed:
Directly define _FILE_OFFSET_BITS instead of including config.h

lambertlulala commented:
@masatake 
So weird, my building environment is the same as [MINGW-packages](https://github.com/msys2/MINGW-packages/blob/7c67d9c46554d10276b2017bcfd0290dfe24a1bf/mingw-w64-universal-ctags-git/PKGBUILD), and it will automatically create config.h.

But another approach to replace the old one is directly define `_FILE_OFFSET_BITS`. On my environment, the `sizeof(off_t)` is 8. And no need to include config.h. Therefore, I push another commit for it.

Passing '0LL' instead of '0L' is due to the argument type of offset for _fseeki64 is defined as `__int64` on Windows.

k-takata commented:
@lambertlulala 
`off_t` is always 32 bits on MSVC. So, I think it's better to define our own off_t.

@masatake 
> Is "the original version of mingw" equal to "MSYS2 / testing (MINGW32)" ?

No. It is not installed in the GHA environment and it is installed in `C:\MinGW` in AppVeyor.
It comes with MSYS 1.0.

(edited)
Currently, we are using the original mingw on the `compiler=mingw` environment on AppVeyor:
https://ci.appveyor.com/project/universalctags/ctags/builds/43598413/job/56nvcd9jmdn28370
It might be a good time to drop support for the original mingw.

lambertlulala commented:
@k-takata 
But `off_t` has been defined in `_mingw_off_t.h` as mentioned in description, that's the reason I manually define `_FILE_OFFSET_BITS`. The inclusion of config.h is also for that. But there might be a better approach?

k-takata commented:
I'm thinking something like this:

```C
// Define readtags' own off_t.
#ifdef _WIN32
typedef long long rt_off_t;
#else
typedef off_t rt_off_t;
#endif
```

and use `rt_off_t` instead of `off_t`.

masatake commented:
@k-takata if we define rt_off_t, I think we don't need to include config.h.
Am I correct?

k-takata commented:
correct

masatake commented:
Let me take over this pull request.

masatake commented:
Merged via #40.
Thank you for your contribution and the discussion.
